### ISSUE TITLE
TEXT array field reference shared dictionary failed

### ISSUE THREAD

**Issue by YanzheL** [heavyai/heavydb#425]
I want to use shared dictionary on TEXT array field, but this SQL command throw an exception 
`Exception: Column raw_dns_log.r_cnames must be a dictionary encoded string`
```sql
CREATE TABLE IF NOT EXISTS raw_dns_log (
    domain      TEXT    NOT NULL    ENCODING DICT(32),
    r_cnames    TEXT[]  ENCODING DICT(32),
    SHARED DICTIONARY   (r_cnames) REFERENCES raw_dns_log(domain)
);
```

The following code doesn't work, either.
```sql
CREATE TABLE IF NOT EXISTS raw_dns_log (
    domain      TEXT    NOT NULL    ENCODING DICT(32),
    r_cnames    TEXT[],
    SHARED DICTIONARY   (r_cnames) REFERENCES raw_dns_log(domain)
);
```

It succeed only If I remove the line `SHARED DICTIONARY       (r_cnames) REFERENCES raw_dns_log(domain)`

So, does it mean that I cannot use shared dictionary on TEXT array field?

I didn't find any information of this constraint in your documentation.

**Comment by YanzheL** [comment]
This doesn't work, either.
```sql
CREATE TABLE IF NOT EXISTS raw_dns_log (
    domain      TEXT    NOT NULL    ENCODING DICT(32),
    r_cnames    TEXT[],
    SHARED DICTIONARY   (r_cnames) REFERENCES raw_dns_log(domain)
);
```

**Comment by randyzwitch** [comment]
If I'm reading this code correctly, your shared dictionary statement is referencing the same table. For the first table, you define the column as `TEXT ENCODING DICT` in your create table statement. 

For future tables defined after this one, if you want to share the dictionary from `raw_dns_log`, then you use the ` SHARED DICTIONARY   (r_cnames) REFERENCES raw_dns_log(domain)` statement

**Comment by YanzheL** [comment]
> If I'm reading this code correctly, your shared dictionary statement is referencing the same table. For the first table, you define the column as `TEXT ENCODING DICT` in your create table statement.
> 
> For future tables defined after this one, if you want to share the dictionary from `raw_dns_log`, then you use the ` SHARED DICTIONARY (r_cnames) REFERENCES raw_dns_log(domain)` statement

May the two code blocks I posted above be misleading.

What I mean is that there is only one table `raw_dns_log` with two fields, `domain TEXT` and `r_cnames TEXT[]`, and I want the `r_cnames` field share the dictionary from `domain` field.

I tried the SQL statement in the first code block I posted above, which contains exactly the same statement `SHARED DICTIONARY (r_cnames) REFERENCES raw_dns_log(domain)` as you said, but it causes `Exception: Column raw_dns_log.r_cnames must be a dictionary encoded string`.

So I reviewed the documentation, and noted that the shared field cannot specify encoding because it uses the same encoding as the referenced field. Then I removed the encoding specification on `r_cnames` field (see the second code block), but it doesn't work, either.

**Comment by randyzwitch** [comment]
Ok, I think I see what you're saying. In this case, these columns are two different types; you're trying to share a dictionary between a string column and an array column. This won't work, as you can think of the dictionary as a map between two types:

`domain: key::integer => value::string`
`r_cnames: key::integer ==> value::text[]`

In the case of `r_cnames`, the key refers to the entire text array, not the individual text elements within the array.

**Comment by YanzheL** [comment]
Now I understood.

But if a TEXT array cannot reuse the dictionary of another TEXT field, then it will cost lots of memory and less efficient, since the actual encoding of `TEXT[]` is indexing the combination of strings, and the strings inside `TEXT[]` have to be reindexed.

So it would be better if there is a way to reuse the dictionary of the element of an array, not just the array itself.

**Comment by randyzwitch** [comment]
Yes, `TEXT[]` is indexing the combination of the strings in the array, but it will add efficiency if the arrays repeat across rows as well. For example, the Tweetmap indexes the Twitter API feed, and in many of those fields, there will be either empty arrays or similar repeated arrays. So you still get some efficiency, even if it's not as efficient as doing it at the cell level.

**Comment by YanzheL** [comment]
So if we can extend the syntax of `SHARED DICTIONARY` to specify the inner cell dictionary of an array field, then we can get efficiency in both senario, the array level and the cell level.

For example,
`SHARED DICTIONARY   (xxx_field_1) REFERENCES xxx_table_2(xxx_field_2) INNER REF xxx_table_3(xxx_field_3)`

**Comment by psilvest** [comment]
This issue has been inactive for a while, and the project has changed substantially. If this still occurs on the latest release, please add reproduction steps and environment details.
